include/aos/common/tools/algorithm.hpp

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_ALGORITHM_HPP_
+#define AOS_ALGORITHM_HPP_
+
+#include "aos/common/tools/error.hpp"
+
+namespace aos {
+
+/**
+ * Algorithm interface.
+ * @tparam T value type.
+ * @tparam I iterator type.
+ * @tparam CI const iterator type.
+ */
+template <typename T, typename I, typename CI>
+class AlgorithmItf {
```

Comment:
I suppose it's better not to create a class for container utils, but keep them as  separate functions.

Reply:
Why? What would be the benefit?

Reply:
In case of a base class approach, it would better utilize CRTP, but with common type aliases added to parent container classes for convenience:  

class Array<T> : public BaseContainerItf< Array<T> > {
    using Val = T;
    using ConstIter = const T*;
    using Iter = T*;
};

class BaseContainerItf< ParentCont > {
   ParentCont* parent() const { return static_cast<const ParentCont*>(this); }
   bool IsEmpty() const { return parent()->Size() == 0; }
   const typename ParentCont::Val& Front() const {return *parent->begin(); }
}

For approach with external functions. Target container should be added as the first argument. eg.
template <class C>
typename const C::Val& Front(const C& cont) { return *cont.begin(); }

Reply:
```c++
typename const C::Val& Front(const C& cont) { return *cont.begin(); }

Front(mList) 
```
Looks weird. I prefer `mList.Front()`.